pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/ReferralSourceService.java

```diff
@@ -0,0 +1,413 @@
+     * @param attributes unused
+     * @param payload request body, if any, as ReferralSourceRequest
+     */
+    public void initialize(String userId, String realm, Map<String, String> attributes, String payload) {
+        this.userId = userId;
+
+        // this class currently only supports RGP realm, since it is the ony realm that supports this feature
+        // recode the RGP constants in this class if that ever changes
+        if (!realm.equalsIgnoreCase(RGP_REALM)) {
+            throw new DsmInternalError("Invalid realm for ReferralSourceService: " + realm);
+        }
+        DDPInstance instance = DDPInstance.getDDPInstance(realm);
+        if (instance == null) {
+            throw new DsmInternalError("Invalid realm: " + realm);
+        }
+        esIndex = instance.getParticipantIndexES();
+        if (StringUtils.isEmpty(esIndex)) {
+            throw new DsmInternalError("No ES participant index for study " + realm);
+        }
+
+        ParticipantDataDao dataDao = new ParticipantDataDao();
+        participantDataByPtpId = new HashMap<>();
+
+        // handle optional list of ptps
+        if (!StringUtils.isBlank(payload)) {
+            List<String> participants = getParticipantList(payload);
+            for (String participantId: participants) {
+                List<ParticipantData> ptpData = dataDao.getParticipantDataByParticipantId(participantId);
+                if (ptpData.isEmpty()) {
+                    throw new DSMBadRequestException("Invalid participant ID: " + participantId);
+                }
+                participantDataByPtpId.put(participantId, ptpData);
+            }
+        } else {
+            // get study participants and their data
+            // Implementation note: we can either gather all the ptp data now (fewer queries, larger memory footprint),
+            // or just gather the ptp IDs here and query for ptp data separately (more queries, smaller memory footprint)
+            // At the time of implementation ptps per study were reasonably small (~100) and ptp data size was not large
+            List<ParticipantData> dataList =
+                    dataDao.getParticipantDataByInstanceId(instance.getDdpInstanceIdAsInt());
+            participantDataByPtpId = dataList.stream().collect(Collectors.groupingBy(ParticipantData::getRequiredDdpParticipantId));
+        }
+    }
+
+    /**
+     * Run the asynchronous part of the operation, updating the AdminRecord with the operation results
+     *
+     * @param operationId ID for reporting results
+     */
+    public void run(int operationId) {
+        List<UpdateLog> updateLog = new ArrayList<>();
+
+        // for each participant attempt an update and log results
+        for (var entry: participantDataByPtpId.entrySet()) {
+            String ddpParticipantId = entry.getKey();
+            try {
+                UpdateStatus status = updateReferralSource(ddpParticipantId, entry.getValue(),
```

Comment:
Just one concern, is this supposed to update referral source for all participants, or the ones that have referral source empty?

Reply:
The logic for that is at ~line 227. (We have to do it there since we have to inspect the existing participant data records.)
But we will update any value that is NA. For any other value, we do not update. So if the CRC assigned a value, or if the survey was filled out recently (since adding support for this feature), it will not be overwritten.